packages/caravan-fees/src/btcTransactionTemplate.ts

```diff
@@ -0,0 +1,686 @@
+   * @throws Error if PSBT parsing fails or required information is missing
+   */
+  static fromPsbt(
+    psbtHex: string,
+    options: Omit<TransactionTemplateOptions, "inputs" | "outputs">,
+  ): BtcTransactionTemplate {
+    const psbt = initializePsbt(psbtHex);
+    const inputs = BtcTransactionTemplate.processInputs(psbt);
+    const outputs = BtcTransactionTemplate.processOutputs(
+      psbt,
+      options.network,
+    );
+
+    return new BtcTransactionTemplate({
+      ...options,
+      inputs,
+      outputs,
+    });
+  }
+
+  /**
+   * Gets the inputs of the transaction.
+   * @returns A read-only array of inputs
+   */
+  get inputs(): readonly BtcTxInputTemplate[] {
+    return this._inputs;
+  }
+
+  /**
+   * Gets the outputs of the transaction.
+   * @returns A read-only array of outputs
+   */
+  get outputs(): readonly BtcTxOutputTemplate[] {
+    return this._outputs;
+  }
+
+  /**
+   * Gets the malleable outputs of the transaction.
+   * Malleable outputs are all those that can have their amount changed, e.g. change outputs.
+   * @returns An array of malleable outputs
+   */
+  get malleableOutputs(): BtcTxOutputTemplate[] {
+    return this._outputs.filter((output) => output.isMalleable);
+  }
+
+  /**
+   * Calculates the target fees to pay based on the estimated size and target fee rate.
+   * @returns {Satoshis} The target fees in satoshis (as a string)
+   */
+  get targetFeesToPay(): Satoshis {
+    return this.targetFees().toString();
+  }
+
+  /**
+   * Calculates the current fee of the transaction.
+   * @returns {Satoshis} The current fee in satoshis (as a string)
+   */
+  get currentFee(): Satoshis {
+    return this.calculateCurrentFee().toString();
+  }
+
+  /**
+   * Checks if the transaction needs a change output.
+   * @returns {boolean} True if there's enough leftover funds for a change output above the dust threshold.
+   */
+  get needsChange(): boolean {
+    const totalInput = this.calculateTotalInputAmount();
+    const totalOutput = this.calculateTotalOutputAmount();
+    const fee = new BigNumber(this.targetFeesToPay);
+    const leftover = totalInput.minus(totalOutput).minus(fee);
+    return leftover.isGreaterThan(this._dustThreshold);
+  }
+
+  /**
+   * Checks if the current fees are sufficient to meet the target fee rate.
+   * @returns True if the fees are paid, false otherwise
+   */
+  areFeesPaid(): boolean {
+    return this.calculateCurrentFee().gte(this.targetFees());
+  }
+
+  /**
+   * Checks if the current fee rate meets or exceeds the target fee rate.
+   * @returns True if the fee rate is satisfied, false otherwise
+   */
+  get feeRateSatisfied(): boolean {
+    return new BigNumber(this.estimatedFeeRate).gte(this._targetFeeRate);
+  }
+
+  /**
+   * Determines if a change output is needed.
+   * @returns True if a change output is needed, false otherwise
+   */
+  get needsChangeOutput(): boolean {
+    const targetFeesWithDustBuffer = this.targetFees().plus(
```

Comment:
I think you might want to add some more buffer. Something tricky with change is that when you add a change output, it makes the tx bigger which increases the fees you need to pay which will reduce the amount that you would add to the change which could make the change dust. In these situations you are actually happy paying a higher fee rate if it would reduce the absolute fee paid by keeping change off and "donating" the extra amount (dust + the buffer) to miners in exchange for the smaller tx without dusty change.

So when determining the buffer, you want to calculate the target fees to pay + the fees to pay for the additional output (rate * output size) + dust. 

It's also kind of ambiguous where we need to do this check. It could be in a fee check or it could be this check if change is needed.

Reply:
would somthing like this work ?

```ts
const MAX_OUTPUT_SIZE = 43; // Large enough to cover P2SH and P2WSH multisig outputs
const changeOutputFee = new BigNumber(this._targetFeeRate).multipliedBy(MAX_OUTPUT_SIZE);

  // Calculate the buffer: target fees + change output fee + dust threshold
  const targetFeesWithBuffer = this.targetFees()
    .plus(changeOutputFee)
    .plus(this._dustThreshold);
```

Reply:
also if we add this then in the RBF-acceleration function in Step:8 

Currently we have :

```ts
 if (newTxTemplate.needsChangeOutput) {
    const totalInputAmount = new BigNumber(newTxTemplate.totalInputAmount);
    const totalOutputAmount = new BigNumber(newTxTemplate.totalOutputAmount);
    const fee = BigNumber.max(
      newTxTemplate.targetFeesToPay,
      txAnalyzer.minimumRBFFee,
    );
    const changeAmount = totalInputAmount.minus(totalOutputAmount).minus(fee);
```

THis will need to be updated to :

```ts
// Step 8: Add or adjust change output
if (newTxTemplate.needsChangeOutput) {
  const changeOutputFee = new BigNumber(targetFeeRate)
    .multipliedBy(MAX_OUTPUT_SIZE)

  const totalInputAmount = new BigNumber(newTxTemplate.getTotalInputAmount());
  const totalOutputAmount = new BigNumber(newTxTemplate.getTotalOutputAmount());
  const fee = BigNumber.max(newTxTemplate.targetFeesToPay, txAnalyzer.minimumRBFFee);
  let changeAmount = totalInputAmount
    .minus(totalOutputAmount)
    .minus(fee)
    .minus(changeOutputFee);
```

as we need to account for the extra fees we will be paying by adding the changes (which the `.needsChangeOutput` checked for)

Reply:
Ahh sorry , I went back to it again and realised that the work to account for the extra fees that was to be added because of adding a new output is resolved by calling the `adjustChangeOutput` fn so we don't need the `changeOutputFee`

> Maybe adjustChangeOutput accounts for this?  (As You described this in the discord )

Reply:
> But I wonder if there's a chance to get in an infinite loop where change is determined to be needed, then you add it, then the new fee amount turns it to dust, it's adjusted and then removed, and then change is determined to be needed.

I don't think this would happen , as this is the only `change(malleable) output` (which is why `needsChangeOutput` returned true in first place) we have in `RBF-Acceleration fn`

So if the output is less than dust , then `adjustChangeOutput` removes it if there are other change outputs in the Tx ... else it just keeps the change as it is with dust amount (to prevent the infinite loop case)

Reply:
> because of adding a new output is resolved by calling the adjustChangeOutput fn so we don't need the changeOutputFee

Right! Exactly. You only really need it to determine if you need to add the output. Once the output is added, all the other tooling that you've added can account for it automatically. It's just this weird in between state when you haven't added the output yet and have to determine if you need it. Without a buffer it would be possible to get stuck in an infinite loop where before you add the output you need a change output, and then when you add the change output the amount becomes dust because more fees are required, so the output is removed which makes it seem like change is needed again, etc.

Reply:
changes [done](https://github.com/caravan-bitcoin/caravan/pull/114/commits/945d644baf5fcfbfac83271deebe440262046414#diff-c113a33821d4e86e4ad4f20743227a51337831792cc5c5926aa2a5971c11d7b1R161)